### DISCUSSION TITLE
apistruct.FullNodeStruct.StateGetReceipt returns nil receipts

### DISCUSSION THREAD

**Discussion by karalabe** [filecoin-project/lotus#5388] [Storage Provider] [answered]
I've tried to interface lotus with its own API implementation from Go. Calling StateGetReceipts with a message CID and the tipset key return nil receipt and nil error. Seems strange, perhaps it's broken?

```go
	tipset, _ := api.ChainHead(context.Background())
	msgs, _ := api.ChainGetBlockMessages(context.Background(), tipset.Blocks()[0].Cid())
	fmt.Println(api.StateGetReceipt(context.Background(), msgs.BlsMessages[0].Cid(), tipset.Key()))
```

Prints `<nil> <nil>`

(Yeah, I dropped all the error handling for this repro, I do check errors locally)

**Comment by magik6k** [comment, marked as answer]
StateGetReceipt looks back for the receipt, and will return nil when the receipt isn't found, which is what happens here, because of the parent execution model (state/receipts it tipset at epoch N are computed using messages/state at epoch N-1)

Random example:
```
$ lotus chain getblock bafy2bzacecgtezx2lvz3mud3jzck2jixewv73yneznmk3tzvfcg5qkbscjs4e | jq '.BlsMessages[0]'
{
  "Version": 0,
  "To": "f065174",
  "From": "f3qg6bktzxmzikb6shagao3p77i5xkq5o6ow7zlf7dx6gzc6hhmndbooq5h4psgyq33zbkdlhtj4wts376my3q",
  "Nonce": 89479,
  "Value": "118278133244029334",
  "GasLimit": 16969083,
  "GasFeeCap": "2946535178",
  "GasPremium": "104705",
  "Method": 6,
  "Params": "igMZvMjYKlgpAAGC4gOB6AIgPICDnbNuL447hbxTe+5uXr+Dl5g9ItB5T4FZxEkpmEsaAAPYJoAaABuUHvQAAAA=",
  "CID": {
    "/": "bafy2bzacebxrcgdz576zrrtxqdrvzgxip7ubw4hedkxk3zcfjhiiy4az34x3y"
  }
}

$ lotus-shed rpc

# Pointing at the block in which the message was included
> StateGetReceipt [{"/":"bafy2bzacebxrcgdz576zrrtxqdrvzgxip7ubw4hedkxk3zcfjhiiy4az34x3y"}, [{"/":"bafy2bzacecgtezx2lvz3mud3jzck2jixewv73yneznmk3tzvfcg5qkbscjs4e"}]]
{"jsonrpc":"2.0","result":null,"id":0}

# Pointing at current chain head, where the message was executed
> StateGetReceipt [{"/":"bafy2bzacebxrcgdz576zrrtxqdrvzgxip7ubw4hedkxk3zcfjhiiy4az34x3y"}, null]
{"jsonrpc":"2.0","result":{"ExitCode":0,"Return":null,"GasUsed":13622267},"id":0}


```

**Comment by karalabe** [comment]
That's a bit odd. Any particular reason for this logic? I'd expect to say "I want the receipt from block X", not "I want the receipt that was executed on top of block X". I'm actually curious why is the block even configurable, isn't there only one "canonical" receipt for any included message?

**Comment by qiukeren** [comment]
Yes this is misunderstanding。

Or i recommend just implement the api handbook https://github.com/filecoin-project/lotus/blob/master/documentation/en/api-methods.md#StateGetReceipt, to clarify the api

🤣🤣🤣🤣🤣🤣🤣🤣🤣🤣
